File: angular/raku/src/app/service/track.service.ts
import {Injectable} from '@angular/core';
import {BaseService} from "./base.service";
import {Observable} from "rxjs";
import {HttpEvent, HttpRequest} from "@angular/common/http";

@Injectable({
  providedIn: 'root'
})
export class TrackService extends BaseService {

  constructor() {
    super();
    this.contextUrl = "/api/track";
  }

  uploadAudio(fileToUpload: File): Observable<HttpEvent<any>> {
    const formData: FormData = new FormData();
    formData.append('file', fileToUpload, fileToUpload.name);
    const req = new HttpRequest('POST', this.BASE_URL + this.contextUrl + "/upload-audio", formData, {
      reportProgress: true,
      responseType: 'json'
    });
    return this.http.request(req);
  }

  create(formData: FormData): Observable<any> {
    return this.http.post(this.BASE_URL + this.contextUrl + "/create", formData);
  }

  update(formData: FormData): Observable<any> {
    return this.http.post(this.BASE_URL + this.contextUrl + "/update", formData);
  }

  getByCode(username, code): Observable<any> {
    return this.http.get(this.BASE_URL + this.contextUrl + "/get", {params: {username: username, code: code}});
  }

  getUserTrackInfo(username, code): Observable<any> {
    return this.http.get(this.BASE_URL + this.contextUrl + "/user-track-info", {
      params: {
        username: username,
        code: code
      }
    });
  }

  getTrackStats(username: string, code: string): Observable<any> {
    return this.http.get(this.BASE_URL + this.contextUrl + "/get-track-stats", {
      params: {
        username: username,
        code: code
      }
    });
  }

  deleteTrack(track): Observable<any> {
    return this.http.get(this.BASE_URL + this.contextUrl + "/delete/" + track.id);
  }

  getTracksOf(username: string): Observable<any> {
    return this.http.get(this.BASE_URL + this.contextUrl + "/get-tracks/" + username);
  }

  getRepostTracksOf(username: string): Observable<any> {
    return this.http.get(this.BASE_URL + this.contextUrl + "/get-repost-tracks/" + username);

  }
}

File: src/main/java/com/nyat/raku/service/TrackService.java
package com.nyat.raku.service;

import com.nyat.raku.model.PlaylistDTO;
import com.nyat.raku.model.TrackDTO;
import com.nyat.raku.payload.TrackStats;
import com.nyat.raku.payload.UserTrackInfo;

import java.util.List;

public interface TrackService {
    TrackDTO create(TrackDTO trackDTO);

    TrackDTO get(Integer id);

    List<TrackDTO> getAll();

    void update(TrackDTO trackDTO);

    void delete(Integer id);

    TrackDTO getByCode(String username, String code) throws Exception;

    UserTrackInfo getUserTrackInfo(String uploader, String code, String username);

    TrackStats getTrackStats(String username, String code) throws Exception;

    List<TrackDTO> getTracksOf(String username);

    List<TrackDTO> getRepostTracksOf(String username);

    List<PlaylistDTO> getPlaylistsOf(String username);
}
